File: backend-dotnet/src/Veha.Api/Services/ReportExport.cs
using ClosedXML.Excel;
using Veha.Api.Dtos;
using Veha.Domain.Enums;

namespace Veha.Api.Services;

/// <summary>Серверная генерация XLSX-отчётов (utilization/таймшиты/портфель) на ClosedXML.
/// Порт report_export.py.</summary>
public static class ReportExport
{
    private const string MoneyFmt = "# ##0.00\\ ₽";
    private const string HoursFmt = "# ##0.00";
    private const string PctFmt = "0.0";

    private static readonly Dictionary<string, string> StageLabels = new()
    {
        ["presale"] = "Пресейл", ["survey"] = "Обследование", ["design"] = "Проектирование",
        ["implementation"] = "Внедрение", ["pilot"] = "Опытная эксплуатация",
        ["support"] = "Поддержка", ["closed"] = "Закрыт",
    };
    private static readonly Dictionary<string, string> TsStatusLabels = new()
    {
        ["draft"] = "Черновик", ["submitted"] = "На утверждении", ["approved"] = "Утверждён", ["rejected"] = "Отклонён",
    };
    private static readonly Dictionary<string, string> RiskLabels = new()
    {
        ["low_margin"] = "Низкая маржа", ["hours_overrun"] = "Перерасход часов",
    };

    private static void Header(IXLWorksheet ws, string[] headers, int row)
    {
        for (var c = 0; c < headers.Length; c++)
        {
            var cell = ws.Cell(row, c + 1);
            cell.Value = headers[c];
            cell.Style.Font.Bold = true;
            cell.Style.Font.FontColor = XLColor.White;
            cell.Style.Fill.BackgroundColor = XLColor.FromHtml("#1F4E78");
        }
    }

    private static byte[] Save(XLWorkbook wb)
    {
        using var ms = new MemoryStream();
        wb.SaveAs(ms);
        return ms.ToArray();
    }

    public static byte[] Utilization(UtilizationReportDto r)
    {
        using var wb = new XLWorkbook();
        var ws = wb.Worksheets.Add("Загрузка ресурсов");
        ws.Cell(1, 1).Value = "Отчёт по загрузке ресурсов (utilization)";
        ws.Cell(1, 1).Style.Font.Bold = true;
        ws.Cell(2, 1).Value = $"Период: {r.DateFrom} — {r.DateTo}  ·  рабочих дней: {r.WorkingDays}  ·  ёмкость на сотрудника: {r.CapacityPerUser} ч";

        Header(ws, ["Сотрудник", "Подразделение", "Билируемые часы", "Ёмкость, ч", "Загрузка, %", "Проектов"], 4);
        var row = 5;
        foreach (var u in r.Rows)
        {
            ws.Cell(row, 1).Value = XlsxSafe.Safe(u.FullName);
            ws.Cell(row, 2).Value = XlsxSafe.Safe(u.Department ?? "—");
            ws.Cell(row, 3).Value = (double)u.BillableHours;
            ws.Cell(row, 4).Value = (double)u.CapacityHours;
            ws.Cell(row, 5).Value = (double)u.UtilizationPct;
            ws.Cell(row, 6).Value = u.ProjectsCount;
            ws.Cell(row, 3).Style.NumberFormat.Format = HoursFmt;
            ws.Cell(row, 4).Style.NumberFormat.Format = HoursFmt;
            ws.Cell(row, 5).Style.NumberFormat.Format = PctFmt;
            row++;
        }
        row++;
        ws.Cell(row, 1).Value = "ИТОГО:";
        ws.Cell(row, 1).Style.Font.Bold = true;
        ws.Cell(row, 3).Value = (double)r.TotalBillableHours;
        ws.Cell(row, 4).Value = (double)r.TotalCapacityHours;
        ws.Cell(row, 5).Value = (double)r.AvgUtilizationPct;
        return Save(wb);
    }

    public static byte[] Timesheets(List<TimesheetExportRow> rows, bool includeCost, DateOnly from, DateOnly to)
    {
        using var wb = new XLWorkbook();
        var ws = wb.Worksheets.Add("Трудозатраты");
        ws.Cell(1, 1).Value = "Выгрузка трудозатрат";
        ws.Cell(1, 1).Style.Font.Bold = true;
        ws.Cell(2, 1).Value = $"Период: {from} — {to}";

        var headers = new List<string> { "Дата", "Сотрудник", "Код", "Проект", "Задача", "Часы", "Статус", "Комментарий" };
        if (includeCost) headers.Add("Себестоимость");
        Header(ws, headers.ToArray(), 4);

        var row = 5;
        decimal totalHours = 0, totalCost = 0;
        foreach (var r in rows)
        {
            totalHours += r.Hours;
            ws.Cell(row, 1).Value = r.WorkDate.ToDateTime(TimeOnly.MinValue);
            ws.Cell(row, 1).Style.DateFormat.Format = "yyyy-mm-dd";
            ws.Cell(row, 2).Value = XlsxSafe.Safe(r.User);
            ws.Cell(row, 3).Value = XlsxSafe.Safe(r.ProjectCode);
            ws.Cell(row, 4).Value = XlsxSafe.Safe(r.ProjectName);
            ws.Cell(row, 5).Value = XlsxSafe.Safe(r.Task ?? "—");
            ws.Cell(row, 6).Value = (double)r.Hours;
            ws.Cell(row, 6).Style.NumberFormat.Format = HoursFmt;
            ws.Cell(row, 7).Value = TsStatusLabels.GetValueOrDefault(r.Status.ToString().ToLowerInvariant(), r.Status.ToString());
            ws.Cell(row, 8).Value = XlsxSafe.Safe(r.Comment);
            if (includeCost)
            {
                totalCost += r.Cost ?? 0m;
                ws.Cell(row, 9).Value = (double)(r.Cost ?? 0m);
                ws.Cell(row, 9).Style.NumberFormat.Format = MoneyFmt;
            }
            row++;
        }
        row++;
        ws.Cell(row, 5).Value = "ИТОГО:";
        ws.Cell(row, 5).Style.Font.Bold = true;
        ws.Cell(row, 6).Value = (double)totalHours;
        ws.Cell(row, 6).Style.NumberFormat.Format = HoursFmt;
        if (includeCost)
        {
            ws.Cell(row, 9).Value = (double)totalCost;
            ws.Cell(row, 9).Style.NumberFormat.Format = MoneyFmt;
        }
        return Save(wb);
    }

    public static byte[] Portfolio(PortfolioResponseDto portfolio, IReadOnlyDictionary<Guid, string> managerNames)
    {
        using var wb = new XLWorkbook();
        var ws = wb.Worksheets.Add("Портфель");
        ws.Cell(1, 1).Value = "Портфель проектов: выручка, маржа, риски";
        ws.Cell(1, 1).Style.Font.Bold = true;
        ws.Cell(2, 1).Value = $"Проектов: {portfolio.ProjectsCount}  ·  выручка: {portfolio.TotalRevenue}  ·  маржа: {portfolio.TotalMargin} ({portfolio.AvgMarginPct}%)  ·  в зоне риска: {portfolio.AtRisk}";

        Header(ws, ["Код", "Проект", "Стадия", "РП", "Выручка", "Маржа", "Маржа, %", "Часы факт", "Часы план", "Перерасход, %", "Риски"], 4);
        var row = 5;
        foreach (var r in portfolio.Rows)
        {
            var risks = r.Risks.Count > 0
                ? string.Join(", ", r.Risks.Select(x => RiskLabels.GetValueOrDefault(x, x)))
                : "—";
            ws.Cell(row, 1).Value = XlsxSafe.Safe(r.Code);
            ws.Cell(row, 2).Value = XlsxSafe.Safe(r.Name);
            ws.Cell(row, 3).Value = StageLabels.GetValueOrDefault(r.Stage.ToString().ToLowerInvariant(), r.Stage.ToString());
            ws.Cell(row, 4).Value = XlsxSafe.Safe(managerNames.GetValueOrDefault(r.ManagerId, "—"));
            ws.Cell(row, 5).Value = (double)r.Revenue;
            ws.Cell(row, 6).Value = (double)r.Margin;
            ws.Cell(row, 7).Value = (double)r.MarginPct;
            ws.Cell(row, 8).Value = (double)r.ActualHours;
            ws.Cell(row, 9).Value = (double)r.PlannedHours;
            ws.Cell(row, 10).Value = (double)r.HoursOverrunPct;
            ws.Cell(row, 11).Value = XlsxSafe.Safe(risks);
            ws.Cell(row, 5).Style.NumberFormat.Format = MoneyFmt;
            ws.Cell(row, 6).Style.NumberFormat.Format = MoneyFmt;
            ws.Cell(row, 7).Style.NumberFormat.Format = PctFmt;
            ws.Cell(row, 8).Style.NumberFormat.Format = HoursFmt;
            ws.Cell(row, 9).Style.NumberFormat.Format = HoursFmt;
            ws.Cell(row, 10).Style.NumberFormat.Format = PctFmt;
            row++;
        }
        return Save(wb);
    }
}

File: backend/app/services/report_export.py
"""Серверная генерация XLSX-отчётов (utilization, таймшиты, портфель) через openpyxl.

Размеры выгрузок в прототипе небольшие — endpoint отдаёт файл синхронно (как и
экспорт ТКП). Для крупных периодов следует вынести в Celery."""

from __future__ import annotations

from datetime import date
from decimal import Decimal
from io import BytesIO

from openpyxl import Workbook
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
from openpyxl.utils import get_column_letter
from openpyxl.worksheet.worksheet import Worksheet

from app.services.xlsx_safe import safe_text

_HEADER_FILL = PatternFill("solid", fgColor="1F4E78")
_HEADER_FONT = Font(color="FFFFFF", bold=True)
_TOTAL_FONT = Font(bold=True)
_THIN = Side(style="thin", color="D9D9D9")
_BORDER = Border(left=_THIN, right=_THIN, top=_THIN, bottom=_THIN)
_MONEY_FMT = r"# ##0.00\ ₽"
_HOURS_FMT = r"# ##0.00"
_PCT_FMT = r"0.0"

_STAGE_LABELS = {
    "presale": "Пресейл",
    "survey": "Обследование",
    "design": "Проектирование",
    "implementation": "Внедрение",
    "pilot": "Опытная эксплуатация",
    "support": "Поддержка",
    "closed": "Закрыт",
}
_TS_STATUS_LABELS = {
    "draft": "Черновик",
    "submitted": "На утверждении",
    "approved": "Утверждён",
    "rejected": "Отклонён",
}
_RISK_LABELS = {"low_margin": "Низкая маржа", "hours_overrun": "Перерасход часов"}


def _write_header(ws: Worksheet, headers: list[str], row: int) -> None:
    for col, title in enumerate(headers, start=1):
        c = ws.cell(row=row, column=col, value=title)
        c.fill = _HEADER_FILL
        c.font = _HEADER_FONT
        c.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)
        c.border = _BORDER


def _autowidth(ws: Worksheet, widths: list[int]) -> None:
    for i, w in enumerate(widths, start=1):
        ws.column_dimensions[get_column_letter(i)].width = w


def build_utilization_xlsx_bytes(report: dict) -> bytes:
    wb = Workbook()
    ws = wb.active
    ws.title = "Загрузка ресурсов"

    ws["A1"] = "Отчёт по загрузке ресурсов (utilization)"
    ws["A1"].font = Font(size=14, bold=True)
    ws["A2"] = (
        f"Период: {report['date_from']} — {report['date_to']}  ·  "
        f"рабочих дней: {report['working_days']}  ·  "
        f"ёмкость на сотрудника: {report['capacity_per_user']} ч"
    )
    ws["A2"].font = Font(italic=True, color="808080")

    headers = [
        "Сотрудник", "Подразделение", "Билируемые часы", "Ёмкость, ч", "Загрузка, %", "Проектов",
    ]
    hr = 4
    _write_header(ws, headers, hr)
    row = hr + 1
    for r in report["rows"]:
        values = [
            r["full_name"],
            r["department"] or "—",
            float(r["billable_hours"]),
            float(r["capacity_hours"]),
            float(r["utilization_pct"]),
            r["projects_count"],
        ]
        for col, val in enumerate(values, start=1):
            c = ws.cell(row=row, column=col, value=safe_text(val))
            c.border = _BORDER
            if col in (3, 4):
                c.number_format = _HOURS_FMT
            elif col == 5:
                c.number_format = _PCT_FMT
        row += 1

    row += 1
    ws.cell(row=row, column=1, value="ИТОГО:").font = _TOTAL_FONT
    tb = ws.cell(row=row, column=3, value=float(report["total_billable_hours"]))
    tb.font = _TOTAL_FONT
    tb.number_format = _HOURS_FMT
    tc = ws.cell(row=row, column=4, value=float(report["total_capacity_hours"]))
    tc.font = _TOTAL_FONT
    tc.number_format = _HOURS_FMT
    ta = ws.cell(row=row, column=5, value=float(report["avg_utilization_pct"]))
    ta.font = _TOTAL_FONT
    ta.number_format = _PCT_FMT

    _autowidth(ws, [32, 22, 18, 14, 14, 12])
    buf = BytesIO()
    wb.save(buf)
    return buf.getvalue()


def build_timesheet_xlsx_bytes(
    rows: list[dict], *, include_cost: bool, date_from: date, date_to: date
) -> bytes:
    wb = Workbook()
    ws = wb.active
    ws.title = "Трудозатраты"

    ws["A1"] = "Выгрузка трудозатрат"
    ws["A1"].font = Font(size=14, bold=True)
    ws["A2"] = f"Период: {date_from} — {date_to}"
    ws["A2"].font = Font(italic=True, color="808080")

    headers = ["Дата", "Сотрудник", "Код", "Проект", "Задача", "Часы", "Статус", "Комментарий"]
    if include_cost:
        headers.append("Себестоимость")
    hr = 4
    _write_header(ws, headers, hr)

    row = hr + 1
    total_hours = Decimal("0")
    total_cost = Decimal("0")
    for r in rows:
        total_hours += r["hours"]
        values = [
            r["work_date"],
            r["user"],
            r["project_code"],
            r["project_name"],
            r["task"] or "—",
            float(r["hours"]),
            _TS_STATUS_LABELS.get(str(r["status"]), str(r["status"])),
            r["comment"],
        ]
        if include_cost:
            total_cost += r.get("cost", Decimal("0"))
            values.append(float(r.get("cost", 0)))
        for col, val in enumerate(values, start=1):
            c = ws.cell(row=row, column=col, value=safe_text(val))
            c.border = _BORDER
            if col == 1:
                c.number_format = "yyyy-mm-dd"
            elif col == 6:
                c.number_format = _HOURS_FMT
            elif include_cost and col == 9:
                c.number_format = _MONEY_FMT
        row += 1

    row += 1
    ws.cell(row=row, column=5, value="ИТОГО:").font = _TOTAL_FONT
    th = ws.cell(row=row, column=6, value=float(total_hours))
    th.font = _TOTAL_FONT
    th.number_format = _HOURS_FMT
    if include_cost:
        tc = ws.cell(row=row, column=9, value=float(total_cost))
        tc.font = _TOTAL_FONT
        tc.number_format = _MONEY_FMT

    widths = [12, 28, 14, 34, 28, 10, 16, 40]
    if include_cost:
        widths.append(16)
    _autowidth(ws, widths)
    buf = BytesIO()
    wb.save(buf)
    return buf.getvalue()


def build_portfolio_xlsx_bytes(portfolio: dict, manager_names: dict) -> bytes:
    wb = Workbook()
    ws = wb.active
    ws.title = "Портфель"

    ws["A1"] = "Портфель проектов: выручка, маржа, риски"
    ws["A1"].font = Font(size=14, bold=True)
    ws["A2"] = (
        f"Проектов: {portfolio['projects_count']}  ·  "
        f"выручка: {portfolio['total_revenue']}  ·  "
        f"маржа: {portfolio['total_margin']} ({portfolio['avg_margin_pct']}%)  ·  "
        f"в зоне риска: {portfolio['at_risk']}"
    )
    ws["A2"].font = Font(italic=True, color="808080")

    headers = [
        "Код", "Проект", "Стадия", "РП", "Выручка", "Маржа", "Маржа, %",
        "Часы факт", "Часы план", "Перерасход, %", "Риски",
    ]
    hr = 4
    _write_header(ws, headers, hr)
    row = hr + 1
    for r in portfolio["rows"]:
        risks = ", ".join(_RISK_LABELS.get(x, x) for x in r["risks"]) or "—"
        values = [
            r["code"],
            r["name"],
            _STAGE_LABELS.get(str(r["stage"]), str(r["stage"])),
            manager_names.get(r["manager_id"], "—"),
            float(r["revenue"]),
            float(r["margin"]),
            float(r["margin_pct"]),
            float(r["actual_hours"]),
            float(r["planned_hours"]),
            float(r["hours_overrun_pct"]),
            risks,
        ]
        for col, val in enumerate(values, start=1):
            c = ws.cell(row=row, column=col, value=safe_text(val))
            c.border = _BORDER
            if col in (5, 6):
                c.number_format = _MONEY_FMT
            elif col in (7, 10):
                c.number_format = _PCT_FMT
            elif col in (8, 9):
                c.number_format = _HOURS_FMT
        row += 1

    _autowidth(ws, [14, 34, 18, 24, 16, 16, 10, 12, 12, 14, 30])
    buf = BytesIO()
    wb.save(buf)
    return buf.getvalue()
